staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -127,29 +128,37 @@ func (wm watchersMap) terminateAll(done func(*cacheWatcher)) {
 }
 
 type indexedWatchers struct {
-	allWatchers   watchersMap
+	allWatchers   map[namespacedName]watchersMap
```

Comment:
I spent some time thinking whether we should do that unconditionally.
Initially, I thought it may be unnecessary overhead in majority of cases, but after looking deeper I think it's fine:
(a) the overhead on registration the watcher (additional level of the map) should be fine (I wasn't really concerned about it)
(b) choosing the watchers that we should pick up:
- previously we were iterating over one (potentially large) map
- with this approach we're iterating over up to four maps (they sum of their sizes <= the map from previous approach)

This seems fine. This PR doesn't address the case of watchers scoped to namespace and/or name if index field is provided, but I think this is fine - we can solve it as a follow up if needed.

[I also see that you mentioned that Lists aren't addressed, but that codepath are independent and their performance doesn't affect e.g. watch throughput (and is not optimized enough either), so let's ignore it.

So I think I'm ok with this PR as as once we add proper tests to it.

Reply:
> This PR doesn't address the case of watchers scoped to namespace and/or name if index field is provided, but I think this is fine - we can solve it as a follow up if needed.

yeah... literally the only index we have in-tree after this change is pods spec.nodeName, which I don't think is an issue in terms of cardinality. I'd probably wait to change / improve the indexing approach until we had examples of use cases for indexing that were expected to be used in combination with namespaced watches